Add support for context.Context to SDK API operation requests

jasdel opened the pull request:
Adds support for context.Context to the SDK by adding `WithContext` methods for each API operation, Paginators and Waiters. e.g `PutObjectWithContext`. This change also adds the ability to provide request functional options to the method calls instead of requiring you to use the `Request` API operation method (e.g `PutObjectRequest`).

The SDK will use Context for cancelation of requests, and retry delays for both requests and waiters. For methods that take a Context, the Context is most not be nil. If the Context is nil a panic will occur.

This change also creates a `aws.Context` interface type which is a copy of Go 1.7's context.Context interface type. This is done to allow Go 1.6 SDK users using to take advantage of the Context while also preventing the SDK from forking its request and retry logic by Go version.

### Go 1.6 and below:
SDK requests on Go 1.6 and 1.5 will assign the Context's `Done()` return to `http.Request.Cancel`. 

### Go 1.6 and above:
SDK requests on Go 1.7 and above will use `http.Request.WithContext` to assign the context to the http.Request. The SDK will also use the Context internally to interrupt request and retry delays. In addition, the value returned by `aws.BackgroundContext` is equal to that returned by `context.Background()`

## Usage Examples
```go
func (c *S3) PutObjectWithContext(ctx aws.Context, input *PutObjectInput, opts ...request.Option)(*PutObjectOutput, error)
```

### New Method:
```go
result, err := svc.PutObjectWithContext(ctx, &s3.PutObjectInput{
    Bucket: aws.String("myBucket"),
    Key: aws.String("myKey"),
    Body: object,
})

// Optionally also add request options
result, err := svc.PutObjectWithContext(ctx, &s3.PutObjectInput{
    Bucket: aws.String("myBucket"),
    Key: aws.String("myKey"),
    Body: object,
}, request.WithLogLevel(aws.LogDebugWithHTTPBody))
```

### Old Method:
```go
req, result := svc.PutObjectRequest(&s3.PutObjectInput{
    Bucket: aws.String("myBucket"),
    Key: aws.String("myKey"),
    Body: object,
})
req.HTTPRequest = req.HTTPRequest.WithContext(ctx)
req.Config.LogLevel = aws.LogLevel(aws.LogDebugWithHTTPBody)
err := req.Send()
```

## Additional Changes
* Adds a `Complete` Request handler list that will get called ever time a request is completed. This includes both success and failure. Complete will only be called once per API operation request.
* `private/waiter` package moved from the private group to `aws/request/waiter` and made publically available.
* Adds Context support to all API operations, Waiters(WaitUntil) and Paginators(Pages) methods.
* Adds Context support for s3manager and s3crypto clients.

Fix #861, #455, #454, #667, #1137

xibz commented on aws/request/request_context_test.go:
```diff
@@ -0,0 +1 @@
+package request
```
?

jasdel replied:
yeah... realized that was missing. Adding tests

xibz commented on aws/request/waiter.go:
```diff
@@ -30,26 +18,97 @@ type WaitAcceptor struct {
+
+// WaiterWithLogger returns a waiter option to set the logger a waiter
+// should use to log warnings and errors to.
+func WaiterWithLogger(logger aws.Logger) WaiterOption {
```
same here

xibz commented on aws/request/waiter.go:
```diff
@@ -30,26 +18,97 @@ type WaitAcceptor struct {
+
+// WaiterWithMaxAttempts returns a waiter option to set the max attempts
+// a waiter should make attempting to find a successful outcome.
+func WaiterWithWaitMaxAttempts(attempts int) WaiterOption {
```
Same here

xibz commented on aws/request/waiter.go:
```diff
@@ -30,26 +18,97 @@ type WaitAcceptor struct {
+
+// WaiterWithDelay returns a waiter option to set the delay strategy
+// to be used by the waiter.
+func WaiterWithDelay(fn WaiterDelay) WaiterOption {
```
`WaiterWithDelay` is misleading. I assumed this would return a waiter with a delay. Perhaps a better name? `WaiterOptionDelay`?

Fugiman replied:
`WithDelayedWaiter`?

jasdel replied:
What do you think about something like `WithWaiterDelay`? I think it mixes the best of both suggestions.

Waiter Option examples:
* `WithWaiterDelay`
* `WithWaiterMaxAttempts`
* `WithWaiterLogger`

Request Option examples:
* `WithRetryer`
* `WithLogLevel`

xibz replied:
Yea, I like that a lot better. :+1:

xibz commented on aws/request/waiter.go:
```diff
@@ -30,26 +18,97 @@ type WaitAcceptor struct {
+// options for each request the waiter makes.
+//
+// Replaces all request options on the waiter.
+func WaiterWithRequestOptions(opts ...Option) WaiterOption {
```
same here

xibz commented on aws/request/waiter.go:
```diff
@@ -103,32 +162,33 @@ func (w *Waiter) Wait() error {
 		}
 
-		time.Sleep(time.Second * time.Duration(w.Delay))
+		err = w.Delay()
```
+1

xibz commented on aws/request/waiter_1_7.go:
```diff
@@ -0,0 +1,34 @@
+//
+//     // TODO this is kind of awkward, can context be merged into waiter?
+//     WaiterWithRequestOptions(WithContext(ctx))
+func WaiterWithConstDelayContext(ctx context.Context, delay time.Duration) WaiterDelay {
```
Same here

xibz commented on aws/request/request_pagination.go:
```diff
@@ -61,8 +90,17 @@ func (r *Request) nextPageTokens() []interface{} {
+//
+// Depercated Use Pagination type for configurable pagination of API operations
+func (r *Request) HasNextPage() bool {
+	return len(r.nextPageTokens()) > 0
```
Perhaps log that a deprecated method is being used

xibz replied:
*deprecated

xibz commented on aws/request/request_pagination.go:
```diff
@@ -61,8 +90,17 @@ func (r *Request) nextPageTokens() []interface{} {
 // page of result data. Call .Send() on this request to execute it.
+//
+// Depercated Use Pagination type for configurable pagination of API operations
 func (r *Request) NextPage() *Request {
```
Same here

xibz replied:
*deprecated

xibz commented on aws/request/request_pagination.go:
```diff
@@ -90,6 +128,8 @@ func (r *Request) NextPage() *Request {
 // return true to keep iterating or false to stop.
+//
+// Depercated Use Pagination type for configurable pagination of API operations
 func (r *Request) EachPage(fn func(data interface{}, isLastPage bool) (shouldContinue bool)) error {
```
Here too

xibz replied:
*deprecated

jasdel commented:
Would it be worth the extra methods to generate __WithContext API methods for go 1.7 in addition, or instead of __Ex?

Waiters/Paginators would need to be in addition probably. API operation method would simplify request handler usage.

```go
func (c *S3) PutObjectWithContext(ctx context.Context, in *S3.PutObjectInput, opts ...request.Option) (*s3.PutObjectOutput, error)
```

jasdel commented:
After discussing this, Think its better to drop the __Ex methods, and replace with __WithContext methods instead. The SDK probably will need to define its own `Context` interface. The SDK's Context interface would be a alias for the Go 1.7 context to allow Go 1.6 users to take advantage of the __WithContext API methods also.

jasdel commented:
Updated methods to be `WithContext` instead of `Ex`. Taking a look at API and Waiter retry delay logic.

jasdel commented on aws/types.go:
```diff
@@ -3,8 +3,39 @@ package aws
 
+// Context is an alias of the Go v1.7 stdlib's context.Context interface.
+// It is represented as a SDK interface to enable you to use the "WithContext"
+// API methods.
```
finish sentence, need to mention it lets you use the WithContext methods when using Go 1.6.

jasdel commented:
Need to cleanup new `request.Pagination` type to have a simpler interface.  `HasNextPage` is still needed for `__Pages` callback func signature.

```go
	for p.Next() {
		page := p.Page()
	}
	err := p.Err()
```

jasdel commented:
Cleanup commits to make reviewing easier.

rhysh commented on aws/request/waiter.go:
```diff
@@ -0,0 +1,248 @@
+
+		// Delay to wait before inspecting the resource again
+		delay := w.Retryer.RetryDelay()
+		if ctx != nil {
```
Part of the contract of the context package is that context.Context values are never nil. From https://golang.org/pkg/context/,

> Do not pass a nil Context, even if a function permits it. Pass context.TODO if you are unsure about which Context to use.

Not checking for nil here (and panicking if the caller incorrectly passes nil) is correct; panicking is allowed in the case of programmer error. Callers without a context.Context from their parent can use `context.Background()` (or the equivalent `aws.BackgroundContext()`).

jasdel replied:
Thanks for pointing this out. I think that makes sense. I'll update these new methods to require a context.

jasdel replied:
I think creating a `aws.BackgroundContext()` like you suggested is the best way to extend the existing functionality without the need for nil contexts. I'll put an update together that includes this and remove the nil support.

rhysh commented on aws/request/waiter.go:
```diff
@@ -0,0 +1,248 @@
+		delay := w.Retryer.RetryDelay()
+		if ctx != nil {
+			select {
+			case <-time.After(delay):
```
This *time.Timer will remain on the runtime's timer heap until it expires, in effect leaking (for some amount of time). Memory used in this way can be significant when timeouts are large or request rates are very high. I recommend using `time.NewTimer` so the timer can be stopped if the context is canceled.

jasdel replied:
This makes sense I can switch this to be a `time.Timer` directly.

rhysh commented on service/acm/api.go:
```diff
@@ -107,6 +108,25 @@ func (c *ACM) AddTagsToCertificate(input *AddTagsToCertificateInput) (*AddTagsTo
+// underlying request's retries may be given sub contexts in the future.
+// See the aws/request package's Request and Option types for configurations.
+//
+// If the context is nil, it will be ignored by the API operation.
```
`context.Background()` (or `context.TODO()`) is the context that callers pass when they want the value to be ignored. They never time out, and contain no `Value`s.

jasdel replied:
Will clean up these lines to be more inline with context contract.

rhysh commented on service/acm/api.go:
```diff
@@ -107,6 +108,25 @@ func (c *ACM) AddTagsToCertificate(input *AddTagsToCertificateInput) (*AddTagsTo
+// If the context is nil, it will be ignored by the API operation.
+func (c *ACM) AddTagsToCertificateWithContext(ctx aws.Context, input *AddTagsToCertificateInput, opts ...request.Option) (*AddTagsToCertificateOutput, error) {
+	req, out := c.AddTagsToCertificateRequest(input)
+	if ctx != nil {
```
Checks like this should not be required. When a context is provided, it should be non-nil (per https://golang.org/pkg/context/).

jasdel replied:
removing these checks.

rhysh commented on service/acm/api.go:
```diff
@@ -107,6 +108,25 @@ func (c *ACM) AddTagsToCertificate(input *AddTagsToCertificateInput) (*AddTagsTo
 	return out, err
 }
 
+// AddTagsToCertificateEx is the same as AddTagsToCertificate with the addition of
```
The method suffix here is incorrect; it should be `AddTagsToCertificateWithContext` rather than `AddTagsToCertificateEx`.

jasdel replied:
Thanks for pointing this out. I missed this during the initial conversion.

jasdel commented:
Thanks for the feedback and suggestions @rhysh. I've updated the PR with an `aws.BackgroundContext()` instead of using nil value for contexts, and corrected the other issues you pointed out.

xibz commented on aws/request/waiter.go:
```diff
@@ -0,0 +1,297 @@
+func (w Waiter) WaitWithContext(ctx aws.Context) error {
+	attempts := 0
+
+	for {
```
Minor, but I feel this may be a little cleaner
```go
for attempts := 1; ; attempts++ {
}
```

jasdel replied:
sure can change that.

xibz commented on aws/corehandlers/handlers.go:
```diff
@@ -156,7 +166,25 @@ var AfterRetryHandler = request.NamedHandler{Name: "core.AfterRetryHandler", Fn:
 		r.RetryDelay = r.RetryRules(r)
-		r.Config.SleepDelay(r.RetryDelay)
+
+		if sleepFn := r.Config.SleepDelay; sleepFn != nil {
```
It seems like if a `SleepDelay` function is specified and someone canceled the request, that `SleepDelay` will always win. Is that what we want?

jasdel replied:
see above

xibz commented on aws/request/waiter.go:
```diff
@@ -0,0 +1,297 @@
+// check the resource state. The passed in attempt is the number of times the
+// Waiter has checked the resource state.
+//
+// Attempt is 1 based, and is the number of times the waiter has attempted
```
What do you mean '1' based?

jasdel replied:
yeah... not really that clear. I'll fix this.  Was supposed to convey that the numbering starts at `1` as in attempts made, like first, second, ect. This is apposed to the "standard" numbering of starting at 0.

xibz commented on aws/request/waiter.go:
```diff
@@ -0,0 +1,297 @@
+
+		// Delay to wait before inspecting the resource again
+		delay := w.Delay(attempts)
+		if sleepFn := req.Config.SleepDelay; sleepFn != nil {
```
Still confused to why we don't care about cancelling here

jasdel replied:
`SleepDelay` is a testing function made available via `aws.Config`.  For backwards compatibility, if the sleep delay function is set, we need to respect it.

Though pointing this out does show that I do need to update the docstsring for `SleepDelay` to state that setting SleepDelay will prevent the SDK using context to interrupt retry delays. Context would still interrupt in flight connections.

xibz commented on aws/request/waiter_test.go:
```diff
@@ -88,26 +97,21 @@ func TestWaiterPathAll(t *testing.T) {
 			{
-				State:    "success",
-				Matcher:  "pathAll",
+				State:    request.SuccessWaiterState,
```
:+1: for constants

rhysh commented on private/model/api/operation.go:
```diff
@@ -139,8 +139,27 @@ func (c *{{ .API.StructName }}) {{ .ExportedName }}Request(` +
+// underlying request's retries may be given sub contexts in the future.
+// See the aws/request package's Request and Option types for configurations.
+//
+// The Context passed in must not be nil or a panic may occur.
```
Since there's a chance of a panic (and the caller is forbidden from passing a nil context), it may be better to 1) say "The Context provided must not be nil", and 2) for aws.request.Request.SetContext to explicitly check for nil and to panic (as net/http.Request.WithContext does). When the panic is reliable the user will learn how to use the API very quickly, rather than "getting lucky" for a while and only encountering the panic once their app is deployed.

(And same for `*PagesWithContext` and `WaitUntil*WithContext`)

jasdel replied:
Thanks clarified this statement for the operations.

jasdel commented on aws/request/request.go:
```diff
@@ -111,6 +110,62 @@ func New(cfg aws.Config, clientInfo metadata.ClientInfo, handlers Handlers,
+// Go 1.7 and above:
+// The underlying http.Request will have the context set to it, creating a
+// shallow copy of the underlying http.Request. The SDK may create sub contexts in the future for each retry attempt.
+func (r *Request) SetContext(ctx aws.Context) {
```
Makes sense, I'll add the check and panic to SetContext.

jasdel commented on aws/request/request_context.go:
```diff
@@ -0,0 +1,16 @@
+import "github.com/aws/aws-sdk-go/aws"
+
+// setContext updates the Request to use the passed in
+// context for cancelization. Context will also be used
```
Fixed spelling issues. thanks

jasdel commented on aws/types.go:
```diff
@@ -3,8 +3,88 @@ package aws
+// This context function is equivilent to context.Background in the Go stdlib.
+//
+// See https://golang.org/pkg/context on how to use contexts.
+func BackgroundContext() Context {
```
I don't see any issue making this change. I'll update Background to source the background based on Go version.

jasdel replied:
The `emptCtx.String` method seems to be primarily used by logging and building a string representation of the [chain of contexts created](https://golang.org/src/context/context.go?#L331). The `golang.org/x/net/context` used the [String method for this well](https://github.com/golang/net/blob/a6577fac2d73be281a500b310739095313165611/context/pre_go17.go#L166).

jasdel commented:
Thanks a lot for the feedback. I've updated the docs clarifying the requirement that Context is non-nil and that a panic will occur if it is. This latest update also makes `SetContext` panic if context is nil creating a stronger contract and expectation.

Cleaned up other documentation throughout the change to clarify functionality and expectations for waiters and paginators. Along with spelling errors in this change.

rhysh commented on aws/request/waiter.go:
```diff
@@ -0,0 +1,298 @@
+			sleepFn(delay)
+		} else {
+			select {
+			case <-time.After(delay):
```
This `time.Timer` will leak for a short period when the context is canceled. It could use `time.NewTimer` as in `AfterRetryHandler`, but with canceling the timer in the `<-ctx.Done()` case rather than calling `defer` since this is in a loop.

jasdel replied:
Thanks, I missed this one. I'll update the waiter and request retry delay flow to share the same logic.

xibz commented:
:shipit:

rhysh commented:
👍
